app/controllers/LongerCommodityCodeController.scala

```diff
@@ -54,61 +55,88 @@ class LongerCommodityCodeController @Inject() (
       }
   }
 
-  def onSubmit(mode: Mode, recordId: String): Action[AnyContent]           =
+  def onSubmit(mode: Mode, recordId: String): Action[AnyContent] =
     (identify andThen getData andThen requireData).async { implicit request =>
-      val commodityCodeOption = request.userAnswers
-        .get(RecordCategorisationsQuery)
-        .flatMap(x => x.records.get(recordId).map(x => x.commodityCode))
+      val categorisationInfoOpt = request.userAnswers.get(RecordCategorisationsQuery).flatMap(_.records.get(recordId))
 
-      commodityCodeOption match {
+      val originalComcodeOpt = categorisationInfoOpt.map(_.originalCommodityCode)
+      val latestComcodeOpt   = categorisationInfoOpt.map(_.commodityCode)
+
+      originalComcodeOpt match {
         case Some(shortCommodity) if commodityCodeSansTrailingZeros(shortCommodity).length == 6 =>
           val shortCode = commodityCodeSansTrailingZeros(shortCommodity)
           form
             .bindFromRequest()
             .fold(
               formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, shortCode, recordId))),
               value => {
-                val longCommodityCode = s"$shortCode$value"
-                (for {
-                  validCommodityCode      <- ottConnector.getCommodityCode(
-                                               longCommodityCode,
-                                               request.eori,
-                                               request.affinityGroup,
-                                               UpdateRecordJourney,
-                                               Some(recordId)
-                                             )
-                  updatedAnswers          <- Future.fromTry(request.userAnswers.set(LongerCommodityCodePage(recordId), value))
-                  updatedAnswersWithQuery <-
-                    Future.fromTry(updatedAnswers.set(LongerCommodityQuery(recordId), validCommodityCode))
-                  _                       <- sessionRepository.set(updatedAnswersWithQuery)
-                } yield Redirect(navigator.nextPage(LongerCommodityCodePage(recordId), mode, updatedAnswers))).recover {
-                  case UpstreamErrorResponse(_, NOT_FOUND, _, _) =>
-                    val formWithApiErrors =
-                      form
-                        .copy(errors = Seq(elems = FormError("value", getMessage("longerCommodityCode.error.invalid"))))
-                    BadRequest(view(formWithApiErrors, mode, shortCode, recordId))
+                val longCommodityCode   = s"$shortCode$value"
+                val shouldRedirectToCya = latestComcodeOpt.exists(latest =>
+                  shortCode.concat(value) == latest &&
+                    categorisationInfoOpt.exists(_.latestDoesNotMatchOriginal)
+                )
+
+                if (shouldRedirectToCya) {
+                  Future.successful(Redirect(routes.CyaCategorisationController.onPageLoad(recordId)))
```

Comment:
Should probably let the navigator handle the actual redirect. Perhaps could pass into the navigator page whether you want it to go to CYA? 

Admittedly seems like a faff for the sake of consistency 😄

Reply:
Sure, will look into it :)